### ISSUE TITLE
Advanced title editor does not open Inkscape on Mac

### ISSUE THREAD

**Issue by domink** [OpenShot/openshot-qt#1789]
**Describe the bug**
I try to edit the titles using the Advanced Editor option at "Edit Title" dialog. But when I click the option, a popup is shown requesting to install inkscape. But Inkscape is installed on my system.
I have opened Open Shot Preferences and changed the content of "Advanced Title Editor (path)" by `/Applications/Inkscape.app/Contents/MacOS/Inkscape` and then Inkscape got opened but empty.

**System Details (please complete the following information):**
- Mac OS 10.13.5
- Hardware: Mac mini model Macmini4,1
   Intel Core 2 Duo (2,4 GHz)
   Caché level 2: 3 MB
   RAM: 8 GB
- OpenShot Version 2.4.2

**To Reproduce**
1. Right click on a title
2. Select "Edit Title" from context menu
3. Click on "Use Advanced Editor"
3.a A popup asks you to install Inkscape
3.b Inkscape gets opened but empty 

**Expected behavior**
Helper information to configure the Advanced Editor

**Screenshots**
Not applicable

**Logs**
[libopenshot.log](https://github.com/OpenShot/openshot-qt/files/2162736/libopenshot.log)
[openshot-qt.log](https://github.com/OpenShot/openshot-qt/files/2162737/openshot-qt.log)

**Additional context**

**Comment by peanutbutterandcrackers** [comment]
@steinybot @DylanC - Could you two please test this out, please? Sorry to bother you. I do not own a mac.

**Comment by peanutbutterandcrackers** [comment]
@domink - Could you please also post a screen-shot of the empty InkScape?

**Comment by domink** [comment]
Sure:
![image](https://user-images.githubusercontent.com/1288430/42271935-e98997d4-7f84-11e8-954a-0f568995ae97.png)
![image](https://user-images.githubusercontent.com/1288430/42271965-fea48c50-7f84-11e8-9b7d-77ab973ba8f2.png)
![image](https://user-images.githubusercontent.com/1288430/42271987-1029aa6e-7f85-11e8-9c2f-83d43acbf941.png)
![image](https://user-images.githubusercontent.com/1288430/42272042-3694a67c-7f85-11e8-9637-cd1d2e1f24a0.png)

Please note that the "Edit Title" dialog remains on top and it is not possible to use the Inkscape app.
Also, I expected that Inkscape title will show the name of the file to edit (temp.svg, I think).
The dialog has all options disabled, the mouse shows a "spinner" and it is not possible to move that dialog.

Attached also the log files for this execution:
[libopenshot.log](https://github.com/OpenShot/openshot-qt/files/2162845/libopenshot.log)
[openshot-qt.log](https://github.com/OpenShot/openshot-qt/files/2162846/openshot-qt.log)

**Comment by peanutbutterandcrackers** [comment]
Hmm... thank you for the logs and screenshots. Hopefully someone else will be able to confirm this soon and we can work on resolving this in the 2.4.2.x bug fix series. :+1:

**Comment by domink** [comment]
I think that I found a workaround for the issue opening inkscape on Mac.
Reading this article http://wiki.inkscape.org/wiki/index.php/Mac_OS_X#Inkscape_command_line, I tried using this on the preferences:
/Applications/Inkscape.app/Contents/Resources/bin/inkscape
And it worked, Inkscape is opened with the selected file (temp.svg) 

But the Open Shot "Edit Title" dialog still on top of all Windows :-(.
Workaround, move the Inkscape instance to another Virtual Desktop and work there.

**Comment by N3WWN** [comment]
I'm glad you found a way to resolve the Inkscape issue.  I was going to suggest putting the `/` in front of the path so that the path is absolute instead of relative to _somewhere_ (though I'm not sure where that somewhere may be off the top of my head).

**Comment by peanutbutterandcrackers** [comment]
@dominik - Awesome! This needs some more testing, IMHO, to mark it as a bug here in OpenShot. I wonder if there's something else going on causing the trouble (InkScape issue? Window Manager issue? I don't really know, though)...

**Comment by domink** [comment]
@peanutbutterandcrackers, it is @domink instead @dominik. (I've found @dominik is also valid but... not sure he will know what we are talking about :-) ).

Please, note that I've found a way to start inkscape from openshot but... the "On Top Always" state of the openshot dialog looks an issue (bug?) which should be handled. Maybe like a separate issue for tracking. What do you think?

**Comment by peanutbutterandcrackers** [comment]
@domink - Oops, Sorry! My bad. lol

Hmm.... I see. Perhaps there is something with that window that is set to a parameter that makes it always on top on a mac? I think this needs more testing.

@ferdnyc - Sir, do you mind taking a very quick look at this, please?

**Comment by ferdnyc** [comment]
> Perhaps there is something with that window that is set to a parameter that makes it always on top on a mac?

No, everything's set up correctly with the window. The problem is how OpenShot itself interacts with Inkscape.

I haven't looked at all of the code yet, but just based on a _user_ side evaluation, when OpenShot launches Inkscape it appears to completely block waiting for it (Inkscape) to exit. And I mean _completely_ block. If I try to resize the Title Editor window while Inkscape is still running, it gets corrupted because OpenShot's event loop isn't being processed to repaint the window contents:

![screenshot from 2018-07-07 02-31-25](https://user-images.githubusercontent.com/538020/42407834-5ea60306-8190-11e8-9b79-d19600162156.png)

Eventually, if I interact with the hung OpenShot for more than a couple of seconds, Gnome Shell notices that it's hung and throws up one of these:

![screenshot from 2018-07-07 02-31-51](https://user-images.githubusercontent.com/538020/42407838-78c0208c-8190-11e8-89bf-b6d3a629a064.png)

Once that message comes up, Inkscape is frozen as well. I suspect that's because it's running as a child of OpenShot. (Of OpenShot's _main_ process, in fact, which is why this is **so** broken.) So when Gnome Shell steps in to manage OpenShot for me, it suspends all of its children in case one of them is a runaway that's causing the hang.

So, I suspect something very similar is happening on MacOS. You can't get rid of the Title Editor window because OpenShot isn't listening to process the instructions from the display server. Inkscape isn't responding because **_OpenShot_ isn't responding**. Honestly, it's a miracle this ever worked anywhere, if indeed it actually did.

There are a few possible solutions, none of them quick or easy:

1. OpenShot starts up a QThread and launches Inkscape from _there_, with the thread hanging until Inkscape exits, and then reporting back to the main thread. This way OpenShot can continue processing its event loop in the main thread, which it absolutely must do even while it's waiting on Inkscape.

1. OpenShot launches Inkscape in a fire-and-forget mode, and _doesn't_ wait for it to exit. It goes right on processing its events, but sets a QFileSystemWatcher on the `.svg` file so it can respond to _any_ updates of the file. (Maybe limit that to only as long as the Title Editor window is open, or something.) OpenShot doesn't care about the lifetime of the Inkscape process, so the user can close it or not, doesn't matter. As long as they save the file when they're done editing it, OpenShot will pick up the changes.

1. OpenShot **always** monitors any/all title files that are loaded into the project, and automatically updates whenever the file changes. Launching Inkscape from OpenShot is merely a convenience function, since the user could also manually launch Inkscape themselves, load and edit the file, and then save it back again to have OpenShot automatically pick up the changes — without even having the Title Editor window open. (In fact, "Advanced Edit..." could even be a context-menu selection on the Project Files panel that launches Inkscape on the title `.svg` without having to open the Title Editor window.)

The third option would also solve #1637, but it has the potential to seriously impact the performance of the entire OpenShot application.

**Comment by peanutbutterandcrackers** [comment]
I smell a PR to fix this cooking up somewhere. *cough* *cough* @ferdnyc *cough* *cough* :laughing:

**Comment by peanutbutterandcrackers** [comment]
Could you please try out the latest daily build and see if this has been resolved? A 'Browse' button has indeed been added to browse to the blender and inkscape executables and that should, hopefully, fix this issue...
